New feature - bookmark all open tabs

pastmaster007 opened the pull request:
**Description**:
New bookmark icon is available top-left corner in tab switcher screen to bookmark all open tabs..\
Fixes #391 

App displays notification toast with saved status of bookmarks\
Added below new tracking parameter for this feature
- "mbm_aot" //MobileBookMark_AllOpenTabs

Removed missing tabs icon from main view controller storyboard\
Enlarged tap area for tab switcher settings icon and bookmark all icon..\
Added english localized messages for user facing messages..\

https://vimeo.com/318577712

**Steps to test this PR**:
1.

1.


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

Commit:
New feature - bookmark all open tabs
New bookmark icon is available top-left corner in tab switcher screen to bookmark all open tabs..
app displays notification toast with saved status of bookmarks
added below new tracking parameter for this feature
- "mbm_aot" //MobileBookMark_AllOpenTabs

Removed missing tabs icon from main view controller storyboard
Enlarged tap area for tab switcher settings icon and bookmark all icon..
Added english localized messages for user facing messages..

bwaresiak commented on DuckDuckGo/Base.lproj/Main.storyboard:
```diff
@@ -69,7 +69,7 @@
                                     </barButtonItem>
                                     <barButtonItem style="plain" systemItem="flexibleSpace" id="hL8-s1-UNT"/>
-                                    <barButtonItem image="Tabs" id="K7o-WL-SC5" userLabel="Tabs Button">
+                                    <barButtonItem title="Tabs" id="K7o-WL-SC5" userLabel="Tabs Button">
```
Nice catch. 😄 Would you mind tweaking it as well for BlankSnapshot storyboard?

pastmaster007 replied:
sure..

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +73,52 @@ class TabSwitcherViewController: UIViewController {
+        }
+    }
+    
+    fileprivate func bookmarkAll(_ openTabs: [Tab] ) -> ([Int], [Int]) {
```
I'd use typealias to make it more obvious what return value is: `typealias BookmarkAllResult = (newBookmarks: [Int], existingBookmarks: [Int])`

This way functions can be tweaked to look like this:
`fileprivate func displayBookmarkAllStatusToast(with result: BookmarkAllResult)`
and
`fileprivate func bookmarkAll(_ openTabs: [Tab] ) -> BookmarkAllResult`

pastmaster007 replied:
nice suggestion for readability..

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +73,52 @@ class TabSwitcherViewController: UIViewController {
     }
     
+    fileprivate func displayBookmarkStatusToast(_ savedBookmarks: [Int], _ alreadySavedBookmarks: [Int]) {
+        if (tabsModel.tabs.count == savedBookmarks.count) || (tabsModel.tabs.count == savedBookmarks.count + alreadySavedBookmarks.count) {
```
Wouldn't just `tabsModel.tabs.count == savedBookmarks.count + alreadySavedBookmarks.count` be enough?

pastmaster007 replied:
thanks for the catch..

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +73,52 @@ class TabSwitcherViewController: UIViewController {
+                    Logger.log(text: "\(link.url) already bookmarked")
+                } else {
+                    bookmarksManager.save(bookmark: link)
+                    savedBookmarks.append(index)
```
Since we are not closing bookmarked tabs, I don't think there a need to keep the record of all new/existing indexes. Maybe just a counters would suffice?

pastmaster007 replied:
for now, i am keeping counters for the logic, in future if we plan to add "bookmarkAll & close" feature, we can revisit..

Commit:
Merge branch 'develop'
removed pixel tracking..

Conflicts:
	DuckDuckGo/Base.lproj/Main.storyboard
	DuckDuckGo/UserText.swift
	DuckDuckGo/en.lproj/Localizable.strings

bwaresiak commented on Core/Pixel.swift:
```diff
@@ -72,7 +72,6 @@ public enum PixelName: String {
     case browsingMenuWhitelist = "mb_wl"
     case browsingMenuReportBrokenSite = "mb_rb"
     case browsingMenuSettings = "mb_st"
-    case browsingMenuFindInPage = "mb_fp"
```
I think this change was unintentional (prolly caused by merge).

pastmaster007 replied:
sorry about that.. it was accidental while removing my pixel tag..

bwaresiak commented on DuckDuckGo/en.lproj/Localizable.strings:
```diff
@@ -151,3 +151,7 @@
+
+"bookmarkAll.tabs.notfound" = "No open tabs found to bookmark";
+"bookmarkAll.tabs.saved" = "All open tabs are bookmarked";
+"bookmarkAll.tabs.failed" = "Failed to bookmark %lu out of %lu tab(s)";
```
I'm wondering about the exact messages that we want to have here - I will come back to you with it in case there's a need for a change before merging PR.

bwaresiak replied:
Let's go with:
- No tabs to bookmark
- All tabs bookmarked
- Unable to bookmark some tabs

This way in the case of the failure (that corresponds to the last message), we don't need to return counts at all, just simple bool flag to indicate the problem.

pastmaster007 replied:
👍🏻

bwaresiak commented on DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift:
```diff
@@ -87,7 +87,7 @@ extension TabViewController {
     
     private func buildFindInPageAction(forLink link: Link) -> UIAlertAction? {
         return UIAlertAction(title: UserText.findInPage, style: .default) { [weak self] _ in
-            Pixel.fire(pixel: .browsingMenuFindInPage)
```
I don't think this was intentional change.

pastmaster007 replied:
you are correct.. it wasn't..

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +73,51 @@ class TabSwitcherViewController: UIViewController {
+        
+        guard tabsModel.tabs.count > 0 else {
+            self.view.showBottomToast(UserText.bookmarkAllTabsNotFound)
+            Logger.log(text: "No open tabs found to bookmark")
```
We probably don't want to log this as this is a regular scenario.

pastmaster007 replied:
sure..

Commit:
added missing pixel tracking from previous checkin..
changed missing tab icon to name in BlankSnapshot storyboard..
updated logic for bookmark resultset
added new icon for bookmarkall..
updated localized messages for bookmarkall feature..

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +75,48 @@ class TabSwitcherViewController: UIViewController {
+        } else {
+            let failedToSaveCount = openTabsCount - results.newBookmarksCount - results.existingBookmarksCount
+            Logger.log(text: "Failed to save \(failedToSaveCount) tabs")
+            self.view.showBottomToast(UserText.bookmarkAllTabsFailedToSave)
```
self is not needed here and in line 80

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -28,10 +28,13 @@ class TabSwitcherViewController: UIViewController {
     @IBOutlet weak var doneButton: UIBarButtonItem!
     @IBOutlet weak var plusButton: UIBarButtonItem!
+    
+    typealias BookmarkAllResult = (newBookmarksCount: Int, existingBookmarksCount: Int)
```
I'd move it to line 26 (above variable declarations).

bwaresiak commented on DuckDuckGo/TabSwitcherViewController.swift:
```diff
@@ -72,6 +75,48 @@ class TabSwitcherViewController: UIViewController {
+    @IBAction func onBookmarkAllOpenTabsPressed(_ sender: UIButton) {
+        
+        guard tabsModel.tabs.count > 0 else {
+            self.view.showBottomToast(UserText.bookmarkAllTabsNotFound)
```
self is not needed here.

bwaresiak reviewed (changes requested):
Thank you for applying all tweaks/improvements. I've just left few additional comments for some minor things, once done please resolve conflicts and I'll merge this.

Thank you!

Commit:
Merge branch 'develop' into feature/bookmarkAll-opentabs

* develop:
  New Feature - quick sharing bookmark from bookmarks screen.. (#409)
  removed as unclear exactly what happens
  upload to test flight
  tweak release lane to match removed beta lane
  clean up Fastlane config and move to automatic and phased release
  bump version and update release notes
  improve onboarding experiment 1 (#419)
  fixes for ipad (#415)

# Conflicts:
#	DuckDuckGo/UserText.swift
#	DuckDuckGo/en.lproj/Localizable.strings

Commit:
removed self while accessing view..
changed type alias location of place in file..

bwaresiak reviewed (approved):
Thank you for your contribution! :)
